pyvisfile/xdmf/__init__.py

```diff
@@ -0,0 +1,1119 @@
+    # NOTE: integer ids taken from ``XdmfAttributeType.hpp``
+    Scalar = 200
+    Vector = 201
+    Tensor = 202
+    Matrix = 203
+    Tensor6 = 204
+    GlobalId = 205
+
+
+class AttributeCenter(enum.Enum):
+    """Center of the attribute stored on the mesh."""
+    # NOTE: integer ids taken from ``XdmfAttributeCenter.hpp``
+    Grid = 100
+    Cell = 101
+    Face = 102
+    Edge = 103
+    Node = 104
+    # NOTE: only for `FiniteElementFunction` attributes, which are not supported
+    Other = 105
+
+
+class Attribute(XdmfElement):
+    """
+    .. automethod:: __init__
+    """
+
+    def __init__(self, *,
+            name: Optional[str] = None,
+            atype: AttributeType = AttributeType.Scalar,
+            acenter: AttributeCenter = AttributeCenter.Node,
+            parent: Optional[Element] = None,
+            ):
+        """
+        :param parent: if provided, *self* is appended to the element.
+        """
+
+        super().__init__(parent, "Attribute", {
+            "Name": name,
+            "Center": acenter.name,
+            "AttributeType": atype.name,
+            })
+
+# }}}
+
+
+# {{{ data item
+
+def _numpy_to_xdmf_number_type(dtype):
+    if dtype.type in (np.int8, np.int16, np.int32, np.int64, np.int):
+        return DataItemNumberType.Int
+    elif dtype.type in (np.uint8, np.uint16, np.uint32, np.uint64, np.uint):
+        return DataItemNumberType.UInt
+    elif dtype.type in (np.float16, np.float32, np.float64, np.float128):
+        return DataItemNumberType.Float
+    else:
+        raise ValueError(f"unsupported dtype: '{dtype}'")
+
+
+def _data_item_format_from_str(text):
+    if not isinstance(text, str):
+        return DataItemFormat.XML
+
+    # NOTE: this handles two cases (not meant to be too smart about it)
+    # 1. if the text is just `some_file.bin` or `some_file.out`, we assume it's
+    #    a custom binary file.
+    # 2. if the text is `some_file.h5:/group/dataset`, we assume it's an HDF5
+    #    file.
+
+    if ":" not in text:
+        return DataItemFormat.Binary
+
+    try:
+        filename, _ = text.split(":")
+    except ValueError as exc:
+        raise ValueError("cannot determine format from text") from exc
+
+    if filename.endswith(".h5"):
+        return DataItemFormat.HDF
+    else:
+        raise ValueError("cannot determine format from text")
+
+
+def _system_to_xdmf_endian():
+    import sys
+    if sys.byteorder == "little":
+        endian = DataItemEndian.Little
+    elif sys.byteorder == "big":
+        endian = DataItemEndian.Big
+    else:
+        endian = DataItemEndian.Native
+
+    return endian
+
+
+class DataItemType(enum.Enum):
+    """Data layout of an item."""
+    Uniform = enum.auto()
+    HyperSlab = enum.auto()
+    Function = enum.auto()
+
+
+class DataItemNumberType(enum.Enum):
+    """Basic number types for an item."""
+    Char = enum.auto()
+    UChar = enum.auto()
+    Int = enum.auto()
+    UInt = enum.auto()
+    Float = enum.auto()
```

Comment:
As above.

Reply:
Didn't really find any numerical values for these either. They seem to be defined [here](https://gitlab.kitware.com/xdmf/xdmf/-/blob/master/core/XdmfArrayType.cpp#L113) and there are some ids [here](https://gitlab.kitware.com/xdmf/xdmf/-/blob/master/core/XdmfArrayType.hpp#L254), but they're all for `type + precision`, while the xml keeps them separate.